File: ios/StakDemo/Components/DonutRing.swift
import SwiftUI

/// A new account's allocation ring (product audit, 2026-09-05): one arc per
/// bucket, clockwise from the top, a small gap between arcs, drawn in the
/// 150-box the authored donut exports (MsDonut / SimDonut) occupy. The demo
/// account keeps those exports. Mirrors android ui/components/DonutRing.kt.
struct DonutRing: View {
	let shares: [CGFloat]
	let colors: [Color]
	let size: CGFloat

	var body: some View {
		let u = figmaUnit
		let stroke = 28 * u
		let gap: CGFloat = shares.count > 1 ? 5 : 0
		Canvas { context, box in
			let center = CGPoint(x: box.width / 2, y: box.height / 2)
			let radius = (min(box.width, box.height) - stroke) / 2
			var start: CGFloat = -90
			for (i, share) in shares.enumerated() {
				let sweep = max(2, 360 * share - gap)
				var arc = Path()
				arc.addArc(
					center: center, radius: radius,
					startAngle: .degrees(start + gap / 2), endAngle: .degrees(start + gap / 2 + sweep),
					clockwise: false
				)
				context.stroke(arc, with: .color(colors[i % max(colors.count, 1)]), style: StrokeStyle(lineWidth: stroke, lineCap: .butt))
				start += 360 * share
			}
		}
		.frame(width: size, height: size)
	}
}

File: android/app/src/main/java/com/stak/demo/ui/components/DonutRing.kt
package com.stak.demo.ui.components

import androidx.compose.foundation.Canvas
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.geometry.Size
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.StrokeCap
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.unit.dp

/**
 * A new account's allocation ring (product audit, 2026-09-05): one arc per
 * bucket, clockwise from the top, a small gap between arcs, drawn in the
 * 150-box the authored donut exports (ms_donut / sim_donut) occupy. The
 * demo account keeps those exports. Mirrors ios DonutRing.swift.
 */
@Composable
internal fun DonutRing(shares: List<Float>, colors: List<Color>, modifier: Modifier) {
	val u = com.stak.demo.ui.onboarding.figmaUnit()
	Canvas(modifier = modifier) {
		val stroke = (28 * u).dp.toPx()
		val inset = stroke / 2
		val gap = if (shares.size > 1) 5f else 0f
		var start = -90f
		shares.forEachIndexed { i, share ->
			val sweep = 360f * share - gap
			drawArc(
				color = colors[i % colors.size],
				startAngle = start + gap / 2,
				sweepAngle = sweep.coerceAtLeast(2f),
				useCenter = false,
				topLeft = Offset(inset, inset),
				size = Size(size.width - stroke, size.height - stroke),
				style = Stroke(width = stroke, cap = StrokeCap.Butt),
			)
			start += 360f * share
		}
	}
}
